File: matlab_files/python_substituted/dat_processor.m
function [node_lines,elem_lines,g,file_identifiers,nset_lines] = dat_processor(fname)
% This function reads the .dat file produced by ABAQUS and reprocesses it
% so that saving the stress and coorinate values used for the following
% calculations is easier.

% Open the .dat file

fid = fopen(fname);
if fid==-1
    error('File not found or permission denied.');
end

% Initialize different line counters: one to store the lines where '*Node'
% appears, one for '*Element', one for '*Nset', one for 'PART INSTANCE'.
% The first three are needed to narrow down the pieces of the .dat file
% from which the node coordinates and element nodes are read, the last one
% is needed to divide the data based on the different model parts. 
% The vectors where the lines containing these keywords are stored are also 
% initialized, and another vector containing the names of each part is 
% initialized too.

line=0;
node_lines = [];
elem_lines = [];
part_lines = [];
nset_lines = [];
node_keyword_counter = 0;
elem_keyword_counter = 0;
part_keyword_counter = 0;
nset_keyword_counter = 0;
file_identifiers = {};

%% Processing
% Define all the necessary key expressions to search for in the file.

node_search_key = '*Node';
element_search_key = '*Element';
part_search_key = 'PART INSTANCE';
region_finisher_key = '*Nset';
input_paragraph_end = 'OPTIONS BEING PROCESSED';

% This cycle runs until the end of the file

while ~feof(fid)
    % Read a line without including the newline
    s=fgets(fid);
    % For every new line in the iteration, check for the different keywords
    node_keyword_flag = contains(s,node_search_key);
    elem_keyword_flag = contains(s,element_search_key);
    part_keyword_flag = contains(s,part_search_key);
    region_finisher_flag = contains(s,region_finisher_key);
    end_keyword_flag = contains(s,input_paragraph_end); 
    % Count the line we are at, which is needed to identify the lines
    % containing the search keys and then used to partition the large dat
    % file and rpocess those parts
    line = line + 1;
    
    % Increase the counter for each key expression and save the line it is
    % found at
    
    if (node_keyword_flag == 1)
        node_keyword_counter = node_keyword_counter + 1;
        node_lines(node_keyword_counter,1) = line;
    end
    if (elem_keyword_flag == 1)
        elem_keyword_counter = elem_keyword_counter + 1;
        elem_lines(elem_keyword_counter,1) = line;
    end
    if (region_finisher_flag == 1)
        nset_keyword_counter = nset_keyword_counter + 1;
        nset_lines(nset_keyword_counter,1) = line;
    end
    % In the case of the part name, store it in a separate vector after
    % manipulating its string to use it as a name for the separate csv
    % files
    if (part_keyword_flag == 1)
        part_keyword_counter = part_keyword_counter + 1;
        part_lines(part_keyword_counter,1) = line;
        file_identifiers{part_keyword_counter} = replace(strip(extractAfter(s,": ")),".","_");
    end
    % We can stop processing before the end of the file; after this key
    % expression there is no data we are interested in 
    if (end_keyword_flag == 1)
        end_line = line;
        break;
    end
end

% Save the line where we stop processing as a delimiter for the last piece
% of file to process, then close the file

node_lines(end+1) = end_line;
fclose(fid);

% Open the file again and scan it. This allows us to access every line as
% an element of a cell array that now the file is.

fid = fopen(fname);
if fid==-1
    error('File not found or permission denied.');
end

% Replace all commas with whitespaces

Data = fileread(fname);
Data = strrep(Data, ',', ' ');

% Some lines begin with "LINE", which is not needed and can be deleted
Data = strrep(Data, 'LINE', ' ');
fid = fopen(fname, 'w');
fwrite(fid, Data, 'char');

% Finally, after writing it again, reopen the file to access it as a cell
% array where each line is a cell. This allows for easy looping and element
% inspection over each line.

fid = fopen(fname);
if fid==-1
    error('File not found or permission denied.');
end
g = textscan(fid,'%s','delimiter','\n');

end

File: python_files_python_3/dat_processor.py
def dat_processor(fname):
    # This function reads the .dat file containing node coordinates and the node labels making up each element and
    # stores the lines where they are defined for the next processing operations. Another .dat file, easier to read
    # for the next scripts, is also saved
    # Open the dat file
    with open(fname, "r+") as read_dat:
        # Read the file as a list of lines
        opened_dat = read_dat.readlines()
        print("Processing the dat file to make it more readable")
        # Initialize counters containing a series of lines in the dat file where we will perform the search for node
        # coordinates and element nodes
        node_lines = []
        elem_lines = []
        part_lines = []
        nset_lines = []
        node_keyword_counter = 0
        elem_keyword_counter = 0
        part_keyword_counter = 0
        nset_keyword_counter = 0
        file_identifiers = []
        # Define the keywords used to search for the lines delimiting the start and end of those regions
        node_search_key = '*Node'
        element_search_key = '*Element'
        part_search_key = 'PART INSTANCE'
        region_finisher_key = '*Nset'
        input_paragraph_end = 'OPTIONS BEING PROCESSED'
        line_counter = 0
        end_line_counter = 0
        # Iterate over each line of the dat file to find whether the keywords or keyphrases are contained in it, and if
        # so, save the number of those lines. Again, extract the names of all of the model parts as they will be used to
        # name the csv files containing the node coordinates and element nodes. The dat file only contains the part
        # values, without being divided into the different regions.
        for line in opened_dat:
            line = line.strip()
            line_counter += 1
            node_keyword_flag = line.find(node_search_key)
            elem_keyword_flag = line.find(element_search_key)
            part_keyword_flag = line.find(part_search_key)
            region_finisher_flag = line.find(region_finisher_key)
            end_keyword_flag = line.find(input_paragraph_end)
            if node_keyword_flag != -1:
                node_keyword_counter = node_keyword_counter + 1
                node_lines.append(line_counter)
            elif elem_keyword_flag != -1:
                elem_keyword_counter = elem_keyword_counter + 1
                elem_lines.append(line_counter)
            elif region_finisher_flag != -1:
                nset_keyword_counter = nset_keyword_counter + 1
                nset_lines.append(line_counter)
            elif part_keyword_flag != -1:
                part_keyword_counter = part_keyword_counter + 1
                part_lines.append(line_counter)
                file_identifiers.append(line.split(": ")[1].rstrip().replace(".", "_"))
            elif end_keyword_flag != -1:
                end_line_counter = line_counter
                break
            else:
                pass
        node_lines.append(end_line_counter)

    # Create a new dat file which will be used in the next steps by replacing commas with spaces and taking away the
    # LINE word at the beginning of some of the dat file lines

    new_earth_dat = fname[0:-4] + "_new.dat"
    with open(fname, "r+") as read_dat:
        with open(new_earth_dat, "w") as new_read_dat:
            for line in read_dat:
                replacements = line.replace(',', ' ').replace('LINE', ' ')
                new_read_dat.write(replacements)

    return node_lines, elem_lines, nset_lines, file_identifiers, new_earth_dat
